Add the ability lock settings.json and bindings.json for plugins

Neko-Box-Coder opened the pull request:
If you have comments in either `settings.json` or `bindings.json`, it is possible for plugins to overwrite the json, and erases all the formatting done by the user.

Adding an option to lock these files to avoid losing formatting.

Related to #3385 (Which we should close), #3302 , #2194

dmaluka commented:
Now I'm inclined to think that adding such options would be a good improvement (although I still think it would good to implement what I suggested in https://github.com/zyedidia/micro/pull/3385#issuecomment-2229630375; but that can be implemented independently, and that is about bindings only, not settings).

But, why limit these options to plugins only? `settings.json` or `bindings.json` are also overwritten by `set` and `bind` commands, so why not let `locksettings` and `lockbindings` also prevent those commands from overwriting them? So that users can be sure that nothing will ever mess with their cherished formatting and comments in their config files.

I.e. `set` and `bind` would modify settings or bindings globally (not just for the current buffer like `setlocal`) but for the current micro session only, i.e. not save them in `settings.json` or `bindings.json`?

dmaluka commented:
And another thought: maybe we should go ahead and make it the _default_ behavior?

Or maybe not just yet, but at least eventually?

So it seems that instead of `locksettings` and `lockbuffers` that prevent overwriting when set to true, it would be nicer to add `savesettings` and `savebindings` that prevent overwriting when set to false? And set them by default to true for now, but maybe eventually to false?

Neko-Box-Coder commented:
> Now I'm inclined to think that adding such options would be a good improvement (although I still think it would good to implement what I suggested in [#3385 (comment)](https://github.com/zyedidia/micro/pull/3385#issuecomment-2229630375); but that can be implemented independently, and that is about bindings only, not settings).

Yeah that can be a separate logic.

> But, why limit these options to plugins only? `settings.json` or `bindings.json` are also overwritten by `set` and `bind` commands, so why not let `locksettings` and `lockbindings` also prevent those commands from overwriting them? So that users can be sure that nothing will ever mess with their cherished formatting and comments in their config files.
> 
> I.e. `set` and `bind` would modify settings or bindings globally (not just for the current buffer like `setlocal`) but for the current micro session only, i.e. not save them in `settings.json` or `bindings.json`?

The main use case of this PR (for me at least) is to remove the concern that any newly installed plugin (although this also kinda applies to the builtin comment plugin) will remove all the formatting + comments. 

I don't mind also applying this for `set` and `bind` as well to be consistent. I would assume if the user is calling `set` or `bind`, the `settings.json` and `bindings.json` are expected to be overwritten, but I guess it wouldn't hurt to apply to these commands as well.

> And another thought: maybe we should go ahead and make it the _default_ behavior?
> 
> Or maybe not just yet, but at least eventually?
> 
> So it seems that instead of `locksettings` and `lockbuffers` that prevent overwriting when set to true, it would be nicer to add `savesettings` and `savebindings` that prevent overwriting when set to false? And set them by default to true for now, but maybe eventually to false?

I am leaning towards not making it a default behavior since I think only a minority of the users would care about the settings and bindings to be formatted/commented, and I reckon the user would normally expect the `set` and `bind` to "just work" instead of having to turn this off manually. As for the names (`locksettings` vs `savesettings`), I guess it depends on what the defaults we decided to be?

dmaluka commented:
> I think only a minority of the users would care about the settings and bindings to be formatted/commented, and I reckon the user would normally expect the `set` and `bind` to "just work" instead of having to turn this off manually.

I'm not sure about that. We've seen quite a few complaints about the existing behavior, although it doesn't prove that they represent the majority of users, but we have no proof of the opposite either.

Anyway, yep, we can keep the default as is for now.



> I don't mind also applying this for `set` and `bind` as well to be consistent. I would assume if the user is calling `set` or `bind`, the `settings.json` and `bindings.json` are expected to be overwritten, but I guess it wouldn't hurt to apply to these commands as well.

Actually, now arguing with myself: perhaps we should not overthink it and should leave `set` and `bind` as they are for now. I don't recall any complaints about this particular aspect of the existing behavior.



> As for the names (`locksettings` vs `savesettings`), I guess it depends on what the defaults we decided to be?

It seems it rather depends on what kind of solution we want: what your PR already implements (i.e. mere protection against insidious plugins, i.e. essentially _an ad-hoc workaround for the unfortunate fact that micro allows this sneaky plugins behavior in the first place_), or a "generic" solution for disabling saving settings to `*.json` when setting them, that covers `set` and `bind` commands as well. For the former, the `lock*` names seem more appropriate. For the latter, the `save*` names seem more appropriate, regardless of their default values.

Now, assuming we want the former kind of solution... Perhaps instead of working around with adding options for blocking the annoying plugin behavior, we should just prevent this behavior unconditionally, once and for all? In the case of `TryBindKey()` with overwrite=false, https://github.com/zyedidia/micro/pull/3385#issuecomment-2229630375 would prevent it. As for `TryBindKey()` with overwrite=true, `SetGlobalOption()` and `SetGlobalOptionNative()`, are there any legitimate use cases for plugins to use them? Are there any plugins that actually use them?

Neko-Box-Coder commented:
Yeah, then we can change the behavior of `TryBindKey()` with `overwrite` to `false` which doesn't write to `bindings.json` but instead just set as default like you said. Most of the plugins don't do set `overwrite` to true so I think it is safe to ignore `overwrite` and treat it as always `false`. (And add `RegisterKeybinding()` for consistent naming)

`SetGlobalOption()` / `SetGlobalOptionNative()` on the other hand is a bit tricky since there are plugins that do use it unfortunately. I personally think we should just treat it as `RegisterGlobalOption()` / `RegisterCommonOption()` and override the default value, there shouldn't be any observable difference. In fact, it should be better since it won't override what the user has set in `settings.json` (and deprecate `SetGlobalOption()`)

I think my take is that if we are changing the behavior, we should change both keybindings and options and not 1 of them. If we are planning to only change 1 of them, we might as well just merge this PR.

Neko-Box-Coder commented:
Also, I think it would be nice to add a (1 time?) warning to `set` and `bind` that it will overwrite and reformat the file to let the user know all the formatting and comments will be gone after executing the command.

dmaluka commented:
> `SetGlobalOption()` / `SetGlobalOptionNative()` on the other hand is a bit tricky since there are plugins that do use it unfortunately. I personally think we should just treat it as `RegisterGlobalOption()` / `RegisterCommonOption()` and override the default value (from any previous `RegisterGlobalOption()` calls), there shouldn't be any observable difference.

Ok, I don't know which plugins are those, but I can imagine that. Yeah, it seems it makes sense to handle it as you suggest.

Neko-Box-Coder commented:
I have just updated the plugin functions to not write to any json. 
I removed `locksettings` but kept `lockbindings` even if it doesn't write to json now.

The reason is if the plugin action is irreversible (i.e. delete a file, edit a buffer, etc...), I want to **explicitly** bind it myself instead of having the plugin bind it behind my back. Having this settings guarantees that even if I pressed a keybind by mistake, I know for sure nothing nasty would happen.

The same logic however can't apply to `locksettings`. The fact that a decent amount of plugins **expects** a setting to be there after calling `RegisterGlobalOption()` would mean that if I do not allow the plugin to do that (i.e. `locksettings = true`), many plugins will break since they expect the option is already there.

Neko-Box-Coder commented:
Just rebased it for the merge conflict. Ready to be mergied if good.

dmaluka commented:
Ok, I see @Neko-Box-Coder just pushed an updated version. But it doesn't make a lot of sense to me.

Neko-Box-Coder commented:
@dmaluka 
I just resolved the conflicts but I don't really remember what was the exact outcome of the conversation, I thought we decided the current implementation was good. Let me re-read the whole thing and get back to you in a minute.

Neko-Box-Coder commented on internal/action/bindings.go:
```diff
@@ -261,9 +261,23 @@ func eventsEqual(e1 Event, e2 Event) bool {
+	if l, ok := config.GlobalSettings["lockbindings"]; ok && l.(bool) {
+		return false, errors.New("bindings is locked by the user")
+	}
+	return TryBindKey(k, v, false, false)
```
But that's no longer the "default" keybinding right? 
I think we should treat plugin registered keybinds and user keybinds the same no? 

So for example if an `unbind` is called, it will use the micro's actual default keybindings. I don't think a mutable default keybinding is a good idea.

As for `reload` in your comment, I am not sure what will happen atm, I assume the bindings will be reconstructed from scratch (default + user keybindings + plugin keybindings) without the unloaded plugin I guess.

Neko-Box-Coder replied:
@dmaluka 
I will do the changes in one go after we get to a conclusion for this thread, since the implementation can change quite a bit depending on what conclusion we arrive to.

Neko-Box-Coder replied:
Yeah you are right, I see what you mean now. Then I will just add a flag to prevent the json write for `SetGlobalOption()` and `SetGlobalOptionNative()`

I got confused when trying to read and differentiate `RegisterGlobalOption()`, `RegisterCommonOption()` and `SetGlobalOption()`.

Then in this case, I will just have `TryBindKey()` but without writing to the json.

Neko-Box-Coder commented on internal/config/settings.go:
```diff
@@ -408,6 +409,16 @@ func RegisterGlobalOptionPlug(pl string, name string, defaultvalue any) error {
+
+// **Deprecated**
+func SetGlobalOptionPlug(option, value string) error {
+	return RegisterGlobalOption(option, value)
```
True, missed it. Thanks I will change it.

Neko-Box-Coder commented on runtime/help/plugins.md:
```diff
@@ -223,13 +228,13 @@ The packages and their contents are listed below (in Go type signatures):
-       given value. Same as using the `> set` command. This will try to convert
-       the value into the proper type for the option. Can return an error if the
-       option name is not valid, or the value can not be converted.
+    - **Deprecated** `SetGlobalOption(option, value string) error`: sets an
```
I will update the docs but if a function simply gets redirected to another function that the plugin can call, the "proxy" function should get deprecated no? Since there's no point to have it anymore.

As said before, I forgot to update the docs. I will update it and mention the user/plugin should call `RegisterGlobalOption` instead.

dmaluka replied:
> but if a function simply gets redirected to another function that the plugin can call

It doesn't. `RegisterGlobalOption` that the plugin can call is a different function, it is implemented via `RegisterGlobalOptionPlug()` which does a different thing than `RegisterGlobalOption()`.

dmaluka commented:
LGTM now.

dmaluka commented on runtime/help/options.md:
```diff
@@ -231,6 +231,12 @@ Here are the available options:
 
     default value: `false`
 
+* `lockbindings`: prevent plugins and lua scripts for binding any keys.
```
s/for/from/ ?

dmytrovoytko commented:
Thanks to all of you! 🙏
